debian/patches/wazo_ari_expose_voicemail_functions

```diff
@@ -524,3 +1262,283 @@ Index: asterisk-16.4.0/rest-api/resources.json
+ #define MAX_NUM_CID_CONTEXTS 10
+ 
+@@ -1175,6 +1179,11 @@ static int vm_msg_forward(const char *fr
+ static int vm_msg_move(const char *mailbox, const char *context, size_t num_msgs, const char *oldfolder, const char *old_msg_ids[], const char *newfolder);
+ static int vm_msg_remove(const char *mailbox, const char *context, size_t num_msgs, const char *folder, const char *msgs[]);
+ static int vm_msg_play(struct ast_channel *chan, const char *mailbox, const char *context, const char *folder, const char *msg_num, ast_vm_msg_play_cb cb);
++static int vm_save_greeting(const char* data, unsigned int length, const char* context, const char* voicemail, const char* greeting, const char* format);
++static int vm_get_greeting(char** data, const char* context, const char* voicemail, const char* greeting, const char* format);
++static int get_vm_greeting_path(char* result_path, const char* context, const char* voicemail, const char* greeting, const char* format);
++static int vm_remove_greeting(const char* context, const char* voicemail, const char* greeting);
++
+ 
+ #ifdef TEST_FRAMEWORK
+ static int vm_test_destroy_user(const char *context, const char *mailbox);
+@@ -15273,6 +15282,9 @@ static const struct ast_vm_functions vm_
+ 	.msg_remove = vm_msg_remove,
+ 	.msg_forward = vm_msg_forward,
+ 	.msg_play = vm_msg_play,
++	.save_greeting = vm_save_greeting,
++	.get_greeting = vm_get_greeting,
++	.remove_greeting = vm_remove_greeting
+ };
+ 
+ static const struct ast_vm_greeter_functions vm_greeter_table = {
+@@ -16836,6 +16848,113 @@ play2_msg_cleanup:
+ 	free_user(vmu);
+ 	return res;
+ }
++static int vm_save_greeting(const char* raw_greeting_audio,
++	unsigned int length,
++	const char* context,
++	const char* voicemail,
++	const char* greeting,
++	const char* format)
++{
++	char path[PATH_MAX];
++
++	int make_path_success = get_vm_greeting_path(path, context, voicemail, greeting, format);
++	if(make_path_success != 0) {
++		return -1;
++	}
++
++	FILE *f = fopen(path, "wb");
++	if(!f) {
++		ast_log(LOG_ERROR, "unable to open file at path %s\n", path);
++		return -2;
++	}
++	int num_bytes_written = fwrite(raw_greeting_audio, length, sizeof(char), f);
++
++	if (num_bytes_written < 0) {
++		ast_log(LOG_ERROR, "unable to write file at path %s\n", path);
++		return -3;
++	}
++
++	fclose(f);
++	return 0;
++}
++
++static int vm_get_greeting(char** raw_greeting_audio,
++	const char* context,
++	const char* voicemail,
++	const char* greeting,
++	const char* format)
++ {
++	char path[PATH_MAX];
++
++	int make_path_success = get_vm_greeting_path(path, context, voicemail, greeting, format);
++	if(make_path_success != 0) {
++		return -1;
++	}
++
++	FILE *f = fopen(path, "r");
++
++	if(!f) {
++		ast_log(LOG_ERROR, "unable to read file at path %s\n", path);
++		return -2;
++	}
++
++	fseek(f, 0, SEEK_END);
++	long f_size = ftell(f);
++	rewind(f);
++	*raw_greeting_audio = ast_calloc(f_size + 1, sizeof(char));
++	if(!*raw_greeting_audio) {
```

Comment:
No, because `raw_greeting_audio` is of type `char **`. Notice the line above assigns a value to the pointee (of type `char*`), thus it is the pointee that must be checked.